packages/core/lib/scope.ts

```diff
@@ -0,0 +1,185 @@
+import BrowseError from "./models/BrowseError";
+import IScope from "./interfaces/IScope";
+
+/**
+ * Recursively find a rule matching the name by walking up the scope/environment inheritance chain
+ * @param {string | Word} name The rule name or Word AST node
+ * @param {Scope} scope The scope to use
+ */
+const resolveModule = (name: { name: string }, scope: IScope ) : any => {
+  const moduleName = name.name;
+
+  if (!scope) {
+    throw new BrowseError({
+      message: `Unknown module '${moduleName}'`,
+      node: name,
+    });
+  }
+
+  if (scope.modules[moduleName]) {
+    return scope.modules[moduleName];
+  }
+
+  return resolveModule(name, scope.parent);
+};
+
+/**
+ * Recursively find a rule matching the name by walking up the scope/environment inheritance chain
+ * @param {string | Word} name The rule name or Word AST node
+ * @param {Scope} scope The scope to use
+ */
+const resolveRule = (name: any , scope: IScope) : any => {
+  // TODO:  name.name.name? dafuq, i need some explanation for this
```

Comment:
lol my bad

the AST is gross and we can improve the naming, but it has to do with:
https://github.com/windsorio/browse/blob/master/packages/parser/lib/index.js#L325-L353

the structure is InitRule -> identifier -> word -> (a string), and each AST node in that chain has a property called `name` which points to the child